### ISSUE TITLE
Registration Token is deprecated / Add support for "Next GitLab Runner Token Architecture"

### ISSUE THREAD

**Issue by BornToBeRoot** [cattle-ops/terraform-aws-gitlab-runner#850]
The current registration token is deprecated an will be removed in GitLab 17
https://docs.gitlab.com/ee/ci/runners/register_runner.html#generate-a-registration-token-deprecated

There is a new authentication token available with GitLab 16 (and in GitLab 15 via feature flag `create_runner_workflow_for_admin`)
https://docs.gitlab.com/ee/ci/runners/register_runner.html#generate-an-authentication-token
https://docs.gitlab.com/ee/architecture/blueprints/runner_tokens/

**Comment by kayman-mk** [comment]
Thanks for pointing this out. We should definitely keep this in mind. I have access to GitLab 15.10 only at the moment. So I can't test anything (not an admin to set the feature flag).

It seems that runners have to be pre-registered (and configured) manually in GitLab. Changes are described here (at the bottom): https://docs.gitlab.com/ee/architecture/blueprints/runner_tokens/

Before installing Gitlab 16.6 (will be release on 2023-11-22) the runners have to be installed via the UI (and the new process) again and use the new tokens. Otherwise they won't process any jobs.

GitLab 17 release: >= Nov 2023

**Comment by itskornbailey** [comment]
does the current code work with the new method for runner registration/authentication?

**Comment by Dylanlan** [comment]
I was struggling to auto-register a new runner on latest versions with Gitlab's new registration process, but seems like I got it working.

Looks like my issue was just that the runner token wasn't getting saved to the `config.toml` file, even though I had specified it in the `gitlab_runner_registration_config.registration_token` input

Not sure if I was doing something wrong, but I'm using latest version `6.5.1` of this module, and Gitlab runner version `16.2.1`. I had created a Gitlab group runner manually in their web UI, to get the new runner token (which starts with `glrt-`), and they had a warning about the new registration process:

![gitlab-new-registration-process](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/82ada91f-a094-448e-be85-b33f7890770b)

After applying the terraform, I saw the agent EC2 instance was created and successfully running, but Gitlab UI would keep saying that the runner was `Never contacted`:

![Screenshot 2023-08-18 at 4 20 07 PM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/6d261ffa-540e-4c11-809d-3c40abd5c6b2)

I SSH'ed into the agent instance, and noticed that the `/etc/gitlab-runner/config.toml` file had a value of `null` for the `token` under `[[runners]]`:

<img width="464" alt="token-was-null" src="https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/c83ce2d1-c251-44da-87d6-bb66cf698627">

I fixed it by just manually updating the token in that file, and the gitlab runner service seemed to auto-detect that change, and then the Gitlab UI showed the runner as `Online`, and it started accepting jobs.

It also took me awhile to find that the old process / token was still available under the 3 dots in the Gitlab UI:

![Screenshot 2023-08-18 at 4 31 25 PM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/624e12d3-772e-458f-aa49-f02f2fd14d9f)

Not sure if others are running into this too, but hope this helps!

**Comment by OpenGLShaders** [comment]
@Dylanlan thanks for the workaround - I had the same issue as you. Using the deprecated process / token I wasn't able to get a runner registered with a group, it would always register with the project instead. Were you able to get a group runner working with the old method?

**Comment by Dylanlan** [comment]
I was able to create a new group runner using the deprecated registration process back on September 6th. I'll try creating another one when I have some time to check if it's changed. I had opted not to use the new registration process yet, because it seemed like every time the agent instance got terminated / recreated, I'd need to apply the manual workaround to save the token again

**Comment by OpenGLShaders** [comment]
@Dylanlan that would be awesome, thanks. My changes aren't persisting either. 

I'd also be curious if you're able to share your configuration for this module, I've tried a couple different settings like: locked_to_project and access_level just in case those had an effect, but nothing has seemed to change it.

**Comment by Dylanlan** [comment]
@OpenGLShaders I was able to create 2 new group runners today, using the deprecated registration process.

Here's what my module config looks like:

```
module "gitlab_runner" {
  source   = "cattle-ops/gitlab-runner/aws"
  version  = "6.5.2"

  gitlab_runner_registration_config = {
    registration_token = var.registration_token
    tag_list           = join(",", var.tags)
    description        = "${var.account_name} gitlab runners #${var.runner_number}"
    locked_to_project  = "false"
    run_untagged       = var.run_untagged
    maximum_timeout    = "86400" # 24 hours
    access_level       = "not_protected"
  }

  aws_region = data.aws_region.current.name

  gitlab_runner_version = "16.3.0"

  environment        = "${var.account_name}-runner${var.runner_number}"
  runners_name       = "${var.account_name}-runners${var.runner_number}"
  runners_gitlab_url = "https://gitlab.com"


  #various other config, eg: vpc_id, instance_type, 
  ...
}
```

Let me know if there's any other specific config values that you want me to check!

Note that I haven't upgraded to latest v7 of this module yet, and I'm using a slightly older Gitlab runner version 16.3.0. I'll try spinning up a new runner on latest versions next week if I have time.

If yours is always registering to a project instead of a group, I'd recommend double checking that the registration token that you're using is definitely coming from the group, and not the project.

ie: Go to `https://gitlab.com/groups/<group-name>/-/runners`, and click the 3-dots to find the registration token:

![Screenshot 2023-11-03 at 5 18 56 PM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/8edf622f-9d81-491a-a0b4-fdc45af71870)

I see that a project's CI/CD settings looks similar for configuring a runner, but will have a different token:

![Screenshot 2023-11-03 at 5 19 09 PM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/7691217/013002ad-3cd4-40b7-b79b-1e447c905be0)

**Comment by OpenGLShaders** [comment]
@Dylanlan thanks for your help! 

Found that even though I had set my registration token variable, it was still defaulting to pulling from a SSM store that someone else had set up. So updated that and we are all set now 😃

**Comment by kayman-mk** [comment]
Released with 7.2.0
